### ISSUE TITLE
Python3 and other forks

### ISSUE THREAD

**Issue by russell** [python-sifter/sifter#2]
Hey,

This is a great project!  I started playing around with it using notmuch, and I'm very impressed.

But I can't use this repository, instead I'm running a fork of a fork, 
1. https://github.com/thperret/sifter/tree/python3
2. https://github.com/edgarholleis/sifter
3. https://github.com/russell/sifter/

So I'm wondering what we need to do to get these change integrated?

**Comment by russell** [comment]
@garyp should I start consolidating things and submitting PR's?

**Comment by garyp** [comment]
Hi @russell. The project I initially wrote sifter for has been on the back burner for many years, so I'm not actively using sifter. I'm happy to hear you have found it useful! I'll gladly merge in PRs if you and others want to submit some of the forked changes floating around.

It'd be good to have a maintainer for the project who is actually using it. If you or anyone else is interested in maintaining sifter then I can either grant additional people commit permissions to this repo or update the README to point at a fork that's being actively developed.

**Comment by russell** [comment]
I'd be happy to take over maintaining, I can't guarantee I'm going to be the most active. But I'll try.

I'm also happy to keep it under your GitHub account, I don't think we will get any benefit from moving it.  If we do i would suggest moving it to a GitHub org.  But let's not bother with that for now if your happy to grant me access.

It would be great if you could make me an admin that might be better, because I don't think contributors can add secrets in GitHub actions.  But I really don't know what you'll be presented with when you add me as a collaborator. 

Also it would be great if i could get edit access to the pypi package.  So i can release new versions. https://pypi.org/user/russell/

**Comment by manfred-kaiser** [comment]
Hello,

i'm also interested in this project and would use it in one of my company projects. So I'm also able to contribute and maintaine and  this project during my work time.

My fork is based on https://github.com/thperret/sifter/tree/python3, so it might not compatible to the other forks.
@russell I you need some features from your fork, I can integrate them in my fork, so we have a clean code base

At this time I'm adding type hints to the project, because most of my company projects are using type hints.
When I'm finished with this task, I would implement missing features, which are present in other forks.

About the Pypi Package at https://pypi.org/project/sifter/
If you want, you can give me permission to upload new packages, so I'm able to update it.

**Comment by russell** [comment]
> i'm also interested in this project and would use it in one of my company projects. So I'm also able to contribute and maintaine and this project during my work time.

that is great news

> My fork is based on https://github.com/thperret/sifter/tree/python3, so it might not compatible to the other forks.
> @russell I you need some features from your fork, I can integrate them in my fork, so we have a clean code base

My changes are pretty minor, fixed a bug, and added a github action for testing.

> At this time I'm adding type hints to the project, because most of my company projects are using type hints.
> When I'm finished with this task, I would implement missing features, which are present in other forks.

Nice, i've been interested to see how type hinting help with python development

> About the Pypi Package at https://pypi.org/project/sifter/
> If you want, you can give me permission to upload new packages, so I'm able to update it.

You and me both.



As I mentioned in my previous message, I think that if we are going to move the project we should move it to it's own org, instead of maintainers it under a personal account.  It makes it much easier as maintainers come and go, we don't have to keep adjusting the URLs and readme's to point at the current version.

I also think maintaining it under @garyp's user would be fine, but it really depends what level of access he is able to bestow on collaborators, I don't want to have to nag him if we need to change some repo configuration.

I just added https://github.com/python-sifter as an org @manfred-kaiser what do you think about using it as a place to host the repo? NB, If you decide to start using it, don't fork the sifter repo into it, contributions aren't tracked in Github on forks.

**Comment by russell** [comment]
Oh @manfred-kaiser , i should have mentioned, I'm using this project to sort all my email, so i'm actively using it, at work too, but in a personal manner.

**Comment by garyp** [comment]
@russell @manfred-kaiser : I've added both of you as collaborators on this repository. I don't mind transferring ownership of this repository to the https://github.com/python-sifter org if that would help with your involvement (as long as I'm made one of the owners of the org). But I'd prefer to see you, or anyone else who is interested, first get actively involved in this project by submitting and reviewing pull requests.

> Also it would be great if i could get edit access to the pypi package. So i can release new versions.

I don't mind dealing with releasing new versions. That's not a huge burden. As mentioned above, I'd like to see someone get actively involved in the project before I grant them permission to push versions to pypi.

> It would be great if you could make me an admin that might be better, because I don't think contributors can add secrets in GitHub actions. But I really don't know what you'll be presented with when you add me as a collaborator.

It looks like the only access level available for personal repositories is "collaborator". Fine-grained permissions are only available in repositories owned by GitHub organizations. That's probably a good reason to move this project to an org, but let's see exactly which permission problems we run into with the collaborator level.

> My fork is based on https://github.com/thperret/sifter/tree/python3, so it might not compatible to the other forks.

It's been a while since I looked at that fork, but I remember having concerns about how the python3 conversion was done. For example, I think that branch modified sifter to allow usage of either `str` or `bytes` for input, which doesn't seem right to me. Perhaps that's required to maintain python2 compatibility? But from a python3 perspective, since sifter is dealing with strings, it should be using solely `str` for everything. From a quick glance, the python3 conversion in https://github.com/russell/sifter (which it looks like was originally from https://github.com/jeffmahoney/sifter) seems like a better approach.

Also, unless someone has a need for python2 support, I'd propose we keep things simpler by dropping python2 and making sifter a python3-only project.

> At this time I'm adding type hints to the project, because most of my company projects are using type hints.

That's really great. It'd be nice to have type hints in sifter.

> My changes are pretty minor, fixed a bug, and added a github action for testing.

I actually also just added a GitHub Action to this repo a couple weeks ago in preparation for accepting some new pull requests :smile: I'd be happy to merge in any improvements you have in your version (I haven't looked at yours yet).

**Comment by manfred-kaiser** [comment]
I think moving to an organization has some advantages like fine grained permissions. Your repository is mentioned on http://sieve.info/libraries but by moving your repository to a new location Github makes redirects to the new location. So this should not be a problem. 

It's ok for me to be a collaborator in your repo. If more people become interested in this project, we can move to an organization.

> > Also it would be great if i could get edit access to the pypi package. So i can release new versions.
> 
> I don't mind dealing with releasing new versions. That's not a huge burden. As mentioned above, I'd like to see someone get actively involved in the project before I grant them permission to push versions to pypi.

We could create a Github action to upload new versions to pypi, when a new release in github is created.

> > My fork is based on https://github.com/thperret/sifter/tree/python3, so it might not compatible to the other forks.
> 
> It's been a while since I looked at that fork, but I remember having concerns about how the python3 conversion was done. For example, I think that branch modified sifter to allow usage of either `str` or `bytes` for input, which doesn't seem right to me. Perhaps that's required to maintain python2 compatibility? But from a python3 perspective, since sifter is dealing with strings, it should be using solely `str` for everything. From a quick glance, the python3 conversion in https://github.com/russell/sifter (which it looks like was originally from https://github.com/jeffmahoney/sifter) seems like a better approach.

Converting "basestring" from Python 2 to 3 will result in (str, bytes). But in your context, only "str" is used, so we can drop "bytes". 


> Also, unless someone has a need for python2 support, I'd propose we keep things simpler by dropping python2 and making sifter a python3-only project.

In my opinion, python 2 support will make things more comlicated. An other reason is, that python 2 is end of life.
I need python 3.6 in some of my projects and full support for type hints is supported since python 3.6.
So I think we should support python 3.6+ and drop support for older python versions.

**Comment by manfred-kaiser** [comment]
I have merged my development branch with russell's master branch: https://github.com/manfred-kaiser/sifter/tree/feature_merge_russell

there are some errors, but I will fix them before continuing with the type hints.

**Comment by manfred-kaiser** [comment]
**Unofficial extensions**

By merging russels fork, I have noticed, that there are extensions implemented, which are unofficial.

- pipe (completely unofficial) usage: pipe the contents of the message to a local command
- rewrite (completely unofficial) usage: rewrite search replace ; search and replace are expected to be regular expressions interpreted by python's re.sub routine

In my opinion we should stay as rfc conform as possible and unofficial extensions should be implemented in a new package.

**Comment by manfred-kaiser** [comment]
I have added type hints in my fork https://github.com/manfred-kaiser/sifter

@russell your fork is not meged at the moment, because I had some problems with unit tests after merging. One unit test fails :-( You can find the merge in my branch https://github.com/manfred-kaiser/sifter/tree/feature_merge_russell

@garyp the master works, but is not compatible with your master. I think we should first merge the pull request from @thperret

**Comment by manfred-kaiser** [comment]
@garyp I have created a pull request for the python3 support and type hints.

I have also added you to the "python-sifter" group as owner. If you want, you can transfer owenership to this group. I think it is easier in an group instead of personal repos.

@garyp Is it possible to add me as additional owner to you Pypi Project https://pypi.org/project/sifter/ so I can release new versions?
My username is: ruzima - https://pypi.org/user/ruzima/

If I want to fix some bugs, it would be faster. I want to integrate it in a tool, which I'm using in my company and I need to release new versions very fast, if there are errors.

Please dont misunsterstand me, but if I'm not able to release new versions to pypi as soon as I have fixed a bug, I have to create my own Package. :-(

**Comment by manfred-kaiser** [comment]
I have reorganized the project.

One major change was the creation of an extension registry and extensions are registered as python entry points.
Using entry points for the extension, it is possible to extend sifter with sieve extensions from other packages.

Now it is possible to write unofficial extensions for sieve in a custom python package.
Those extensions are defined as entry point in `setup.py`:

It is important to use the `sifter_extensions` entry point!

Entry points from the sifter package:

```python
    entry_points={
        'sifter_extensions': [
            # sifter commands
            'discard = sifter.commands.discard:CommandDiscard',
            'fileinto = sifter.commands.fileinto:CommandFileInto',
            'if = sifter.commands.if_cmd:CommandIf',
            'elseif = sifter.commands.if_cmd:CommandElsIf',
            'else = sifter.commands.if_cmd:CommandElse',
            'keep = sifter.commands.keep:CommandKeep',
            'redirect = sifter.commands.redirect:CommandRedirect',
            'require = sifter.commands.require:CommandRequire',
            'stop = sifter.commands.stop:CommandStop',
            # sifter tests
            'address = sifter.tests.address:TestAddress',
            'allof = sifter.tests.allof:TestAllOf',
            'anyof = sifter.tests.anyof:TestAnyOf',
            'exists = sifter.tests.exists:TestExists',
            'header = sifter.tests.header:TestHeader',
            'false = sifter.tests.false:TestFalse',
            'not_test = sifter.tests.not_test:TestNot',
            'size = sifter.tests.size:TestSize',
            'true = sifter.tests.true:TestTrue',
            # sifter comparators
            'ascii_casemap = sifter.comparators.ascii_casemap:ComparatorASCIICasemap',
            'octed = sifter.comparators.octet:ComparatorOctet'
        ],
    }
```

Each entry point is loading during parser creation. So there is no need to register extensions in code.

If you need to regiser extensions in your code, this is still possible.
Registering extensions as entry point is recommended, but if are using a single file instead of a package you need to register your extension in the extension registry.

Register a comparator, command, test by using `ExtensionRegistry.register_handler`:

```python
from sifter.extensions import ExtensionRegistry
from sifter.grammar.comparator import Comparator

class MockComparator(Comparator):
    COMPARATOR_ID = 'i;vnd-mock'

ExtensionRegistry.register_handler(MockComparator)
```

Register an extension by using `ExtensionRegistry.register_extension`:

```python
from sifter.extensions import ExtensionRegistry
ExtensionRegistry.register_extension('ext1')
```

**Comment by garyp** [comment]
I agree we should put unofficial extensions into separate python packages. And thanks for the reorg @manfred-kaiser! I had a similar idea of using entry points to auto-discover extensions. I'm glad to see it implemented.

**Comment by garyp** [comment]
I'll transfer this over to the GitHub org and sort out PyPi permissions in the next few days. I've got a lot of other work I'm dealing with right now, so it probably won't be until this weekend.

**Comment by russell** [comment]
@manfred-kaiser i made it about 1/4 of the way through reviewing your branch before you closed the PR.  It's pretty easy to review since it's mostly type changes.  I would really like to see it merged into master, so that additional features are easier to understand.  I'll switch over to using your branch tonight and see if i come up with any feedback of value

**Comment by russell** [comment]
@garyp i made you an owner of the org, i can't remember if i sent you an invitation when i created it. sorry

**Comment by manfred-kaiser** [comment]
What do you think about creating a new Package for Python 3 and use https://pypi.org/project/sifter/ for Python 2 compatibility?

Python 2 is end of life but in some use cases it make sense to have a python 2 version.

**Comment by garyp** [comment]
> What do you think about creating a new Package for Python 3 and use https://pypi.org/project/sifter/ for Python 2 compatibility?
> 
> Python 2 is end of life but in some use cases it make sense to have a python 2 version.

Unless the python2 version is going to be actively maintained, I'd say anyone that needs a python2 version can just fetch the last release of sifter that supported python2.

**Comment by russell** [comment]
@garyp I have sent you an invitation to the org you'll obviously be an owner of the org, I have made @manfred-kaiser you and i all owners, my email is on the billing address, but i doubt we will exceed the free quota https://github.com/python-sifter

**Comment by garyp** [comment]
@russell Thanks for the invitation! I've joined the org but looks like you still need to make me an owner. Let me know when that's done so I can transfer this repo over.

@russell @manfred-kaiser I've added you both as maintainers to the PyPi package.

**Comment by russell** [comment]
> @russell Thanks for the invitation! I've joined the org but looks like you still need to make me an owner. Let me know when that's done so I can transfer this repo over.

@garyp can you try again, i could see you are an owner when i went in, but it could have been lag or some other github quirk.  Anyway, choose to make you an owner again, perhaps it will work better this time

**Comment by manfred-kaiser** [comment]
Currently I'm using another package name, because I needed a package for my own projects, and setting up a private pypi repo for a single package seemed overkill. This is the reason, why I have created the package sifter3 (https://pypi.org/project/sifter3/) 

Looking at the download statistics, there are more than 750 other downloads (in 9 days). Switching development back to sifter might break their projects or leave them without updates :-(

My next steps are:

- switching from ply to sly, because of it is a modernized version of ply and the more up to date release (sly is from 2020 and ply is from 2018) (it is from the same author)
- reimplement the evaluation state to be able to set parser restrictions and extensions can add there own state to the global evaluation state. At this time the states of extensions are hardcoded, which makes it hard or impossible to implement extensions in other python packages.

Due to this changes, a new version might be incompatible with the current version of sifter because switching to sly results in a complete rewrite of the lexer (lexer.py) and parser (grammar.py) Perhaps this involves also changes to other classes. My latest release (sifter3) uses some concepts of sly (lexer and parser as objects), so there are no compatibility problems and switching to sly does not break existing projects.

@russell @garyp So, how should we continue?

**Comment by garyp** [comment]
I've transferred the repository over to the `python-sifter` org. I'm interested in continuing to be involved in this project and I'll provide feedback as I have time available. But please don't wait on my approval for changes to the project, especially if there is consensus among both of you. I'm just glad to see someone else finding it useful and spending time to improve it.

> Currently I'm using another package name, because I needed a package for my own projects, and setting up a private pypi repo for a single package seemed overkill. This is the reason, why I have created the package sifter3 (https://pypi.org/project/sifter3/)

@manfred-kaiser I've added both you and @russell as maintainers to the PyPi project. I would advocate for continuing to publish releases under the `sifter` package name. But if you want to use the `sifter3` package name instead, that's your call :man_shrugging: 

> Looking at the download statistics, there are more than 750 other downloads (in 9 days). Switching development back to sifter might break their projects or leave them without updates :-(

Since the package was only published a week ago, I'd be _very_ surprised if those 750 downloads are anything other than bots and your own usage. The Sieve protocol is a somewhat obscure part of the Internet's tech stack. There just aren't that many people actively developing with it that a new library would suddenly have that many active dependencies on it in a week's time.

> My next steps are:
> 
> * switching from ply to sly, because of it is a modernized version of ply and the more up to date release (sly is from 2020 and ply is from 2018) (it is from the same author)
> * reimplement the evaluation state to be able to set parser restrictions and extensions can add there own state to the global evaluation state. At this time the states of extensions are hardcoded, which makes it hard or impossible to implement extensions in other python packages.

:+1: 

> Due to this changes, a new version might be incompatible with the current version of sifter because switching to sly results in a complete rewrite of the lexer (lexer.py) and parser (grammar.py) Perhaps this involves also changes to other classes. My latest release (sifter3) uses some concepts of sly (lexer and parser as objects), so there are no compatibility problems and switching to sly does not break existing projects.

I wouldn't expect changes to the lexer and parser to cause incompatible changes to sifter's public API, since the lexer and parse are both implementation details. But if you do need to make incompatible changes for whatever reason, I don't see a problem with that as long we're using semantic versioning and making it clear that this is a breaking change. The project is still young enough (in usage, if not in years :smile:) that I think it's fine to make breaking changes.

**Comment by russell** [comment]
@manfred-kaiser 

> Currently I'm using another package name, because I needed a package for my own projects, and setting up a private pypi repo for a single package seemed overkill. This is the reason, why I have created the package sifter3 (https://pypi.org/project/sifter3/)
> 
> Looking at the download statistics, there are more than 750 other downloads (in 9 days). Switching development back to sifter might break their projects or leave them without updates :-(

It's easy for them to switch, so i don't see that as an issue. If you want you could release a new major version that fails to install and advises them to change packages back to sifter. Or prints a deprecation warning if you want to be nicer about it.

it's up to you regarding the naming, i would say keeping the same name as the repo is helpful,  so it would be best to either keep the name sifter, or rename the entire project to sifter3 given that.  I would think keeping the name sifter is easier.

> My next steps are:
> 
> * switching from ply to sly, because of it is a modernized version of ply and the more up to date release (sly is from 2020 and ply is from 2018) (it is from the same author)
> * reimplement the evaluation state to be able to set parser restrictions and extensions can add there own state to the global evaluation state. At this time the states of extensions are hardcoded, which makes it hard or impossible to implement extensions in other python packages.
> 
> Due to this changes, a new version might be incompatible with the current version of sifter because switching to sly results in a complete rewrite of the lexer (lexer.py) and parser (grammar.py) Perhaps this involves also changes to other classes. My latest release (sifter3) uses some concepts of sly (lexer and parser as objects), so there are no compatibility problems and switching to sly does not break existing projects.
> 
> @russell @garyp So, how should we continue?

nice, i would say just release it under a new version, but i don't expect that we would see many complaints, the bit that i have used it, i never touched the lexer or parser directly.

**Comment by manfred-kaiser** [comment]
I have moved my repository of sifter3 to the python-sifter group, because it was not connected to the old sifter project.

I will also keep the new name "sifter3" for the python3 version.

So we can close this issue ticket and the pull request.
